libretro-rs/src/lib.rs

```diff
@@ -110,6 +110,10 @@ impl TryFrom<u32> for RetroDevice {
 pub struct RetroDevicePort(u32);
 
 impl RetroDevicePort {
+  pub fn new(port_num: u32) -> Self {
+    RetroDevicePort(port_num)
+  }
```

Comment:
I looked briefly, but couldn't find if there's an upper limit on what a device port number can be? I'd like to enforce things like this if at all possible. Do you happen to know if there is one? The basic pattern would then be:

```rust
pub const UPPER_LIMIT: u32 = ???;

pub fn from_raw(val: u32) -> Option<Self> {
  if val < Self::UPPER_LIMIT {
    None
  } else {
    Some(Self(val))
  }
}
```

In lieu of an official number, perhaps adding a sufficiently large limit and getting the pattern in place ahead of time is the correct move?

Reply:
Since you asked, I searched up and down libretro.h and also did some searches on the Libretro Discord and came out empty-handed.

To give a sense of scale, RetroArch defaults to 5 ports but can be changed in the UI settings to go as low as 1 or as high as 16. The latter is currently a hard-coded limit; [there's a widely-used constant for it in the code](https://github.com/libretro/RetroArch/blob/b403f5fc1fb78885c8898cf9b94f521dd4cfb70d/input/input_defines.h#L27). But that's just one frontend; others could have different limits. Case in point, [I just found this frontend called Ludo that seems to have a hard limit of 5](https://github.com/libretro/ludo/blob/master/input/input.go). A core ought to function properly regardless of which frontend is driving it.

So as best as I can tell the expectation is that 1) cores will self-impose some limit that makes sense for their use case, 2) cores won't make assumptions about the number of ports a frontend provides, 3) frontends won't make assumptions about the number of ports the core will try to query and 4) cores will gracefully handle ports being set to `RETRO_DEVICE_NONE`.

My 2 cents is that I can't imagine a use case for more than 16 ports but I'd be uncomfortable with this crate setting limits that aren't part of the libretro spec just as a matter of principle. If someone really wants to do that they shouldn't be forced to abandon the crate or code in C/C++.

Reply:
I'm fine with imposing a 16 limit for now. I'm hard pressed to think of games that support more than 4 or 8 in local coop, so 16 seems reasonable.

Reply:
I still think it's a mistake to set the limit at 16; it's an implementation detail that RetroArch could change at any time. The libretro API doesn't provide any way to determine what the frontend's limit is; that implies frontends need to be able to gracefully handle port numbers beyond their self-imposed limit. Case in point, I modified my core to poll gamepad button A on port 17 and RetroArch handled it just fine.

Conversely, RetroArch doesn't always provide 16 ports; it defaults to 5 and could be configured to use as few as 1. Any multiplayer core needs to handle frontends with fewer ports than the core's maximum. A 20 port core would still need to be written to handle as few as 1 port.

That said, I did find that at least on x64 Windows, RetroArch has a hard limit of 65601. I'm able to poll up to port 65600, but polling port 65601 causes a segfault. 65601 is also `u16::MAX + 66`, so we should probably limit the port number to `u16`.

Reply:
I looked further into how [Ludo](https://github.com/libretro/ludo) (which seems to be the main alternative to RetroArch) handles input and [it also seems to handle polling ports beyond their maximum gracefully](https://github.com/libretro/ludo/blob/819878fdbb872d99f5432c87ea0b1aff3447edf5/input/input.go#L177):
```go
if port >= MaxPlayers {
    return 0
}
```